File: clients/ios/BuildIt/Modules/Marketplace/MarketplaceModule.swift
// MarketplaceModule.swift
// BuildIt - Decentralized Mesh Communication
//
// Marketplace module for listings, co-ops, skill exchanges, and resource sharing.

import Foundation
import SwiftUI
import os.log

/// Marketplace module implementation
@MainActor
public final class MarketplaceModule: BuildItModule {
    // MARK: - Module Metadata

    public static let identifier = "marketplace"
    public static let version = "1.0.0"
    public static let dependencies: [String] = []

    // MARK: - Properties

    private let store: MarketplaceStore
    private let service: MarketplaceService
    private let configManager = ModuleConfigurationManager.shared
    private let logger = Logger(subsystem: "com.buildit", category: "MarketplaceModule")

    // MARK: - Initialization

    public init() throws {
        self.store = try MarketplaceStore()
        self.service = MarketplaceService(store: store)
        logger.info("Marketplace module created")
    }

    // MARK: - BuildItModule Implementation

    public func initialize() async throws {
        logger.info("Initializing Marketplace module")

        // Enable by default for global scope
        try await enable(for: nil)

        logger.info("Marketplace module initialized")
    }

    public func handleEvent(_ event: NostrEvent) async {
        // Route marketplace-related Nostr events to service
        await service.processNostrEvent(event)
    }

    public func getViews() -> [ModuleView] {
        [
            ModuleView(
                id: "marketplace",
                title: "Marketplace",
                icon: "storefront",
                order: 35
            ) {
                MarketplaceListView(service: service)
            }
        ]
    }

    public func cleanup() async {
        logger.info("Cleaning up Marketplace module")
    }

    public func isEnabled(for groupId: String?) -> Bool {
        configManager.isModuleEnabled(Self.identifier, for: groupId)
    }

    public func enable(for groupId: String?) async throws {
        configManager.enableModule(Self.identifier, for: groupId)
        logger.info("Enabled Marketplace module for group: \(groupId ?? "global")")
    }

    public func disable(for groupId: String?) async {
        configManager.disableModule(Self.identifier, for: groupId)
        logger.info("Disabled Marketplace module for group: \(groupId ?? "global")")
    }

    // MARK: - Public API

    /// Create a new listing
    public func createListing(
        type: ListingType,
        title: String,
        description: String? = nil,
        price: Double? = nil,
        currency: String = "USD",
        images: [String] = [],
        location: LocationValue? = nil,
        availability: String? = nil,
        tags: [String] = [],
        expiresAt: Date? = nil,
        groupId: String? = nil,
        coopId: String? = nil,
        contactMethod: String = "dm"
    ) async throws -> Listing {
        try await service.createListing(
            type: type,
            title: title,
            description: description,
            price: price,
            currency: currency,
            images: images,
            location: location,
            availability: availability,
            tags: tags,
            expiresAt: expiresAt,
            groupId: groupId,
            coopId: coopId,
            contactMethod: contactMethod
        )
    }

    /// Get all listings
    public func getListings(
        groupId: String? = nil,
        type: ListingType? = nil,
        activeOnly: Bool = false
    ) async throws -> [Listing] {
        try store.getListings(groupId: groupId, type: type, activeOnly: activeOnly)
    }

    /// Delete a listing
    public func deleteListing(_ listingId: String) async throws {
        try store.deleteListing(listingId)
    }

    /// Register a co-op profile
    public func registerCoop(
        name: String,
        description: String? = nil,
        memberCount: Int = 1,
        governanceModel: GovernanceModel = .consensus,
        industry: String = "",
        location: LocationValue? = nil,
        website: String? = nil,
        nostrPubkey: String,
        groupId: String? = nil
    ) async throws -> CoopProfile {
        try await service.registerCoop(
            name: name,
            description: description,
            memberCount: memberCount,
            governanceModel: governanceModel,
            industry: industry,
            location: location,
            website: website,
            nostrPubkey: nostrPubkey,
            groupId: groupId
        )
    }

    /// Get all co-op profiles
    public func getCoopProfiles(groupId: String? = nil) async throws -> [CoopProfile] {
        try store.getCoopProfiles(groupId: groupId)
    }
}

// MARK: - Marketplace Service

/// Service layer for marketplace business logic
@MainActor
public final class MarketplaceService {
    private let store: MarketplaceStore
    private let logger = Logger(subsystem: "com.buildit", category: "MarketplaceService")

    public init(store: MarketplaceStore) {
        self.store = store
    }

    /// Process an incoming Nostr event
    public func processNostrEvent(_ event: NostrEvent) async {
        switch event.kind {
        case MarketplaceNostrKinds.listing:
            await handleListingEvent(event)
        case MarketplaceNostrKinds.coopProfile:
            await handleCoopEvent(event)
        case MarketplaceNostrKinds.review:
            await handleReviewEvent(event)
        case MarketplaceNostrKinds.skillExchange:
            await handleSkillExchangeEvent(event)
        case MarketplaceNostrKinds.resourceShare:
            await handleResourceShareEvent(event)
        default:
            break
        }
    }

    /// Create a new listing
    public func createListing(
        type: ListingType,
        title: String,
        description: String?,
        price: Double?,
        currency: String,
        images: [String],
        location: LocationValue?,
        availability: String?,
        tags: [String],
        expiresAt: Date?,
        groupId: String?,
        coopId: String?,
        contactMethod: String
    ) async throws -> Listing {
        let listing = Listing(
            type: type,
            title: title,
            description: description,
            price: price,
            currency: currency,
            images: images,
            location: location,
            availability: availability,
            tags: tags,
            createdBy: "current-user", // Would come from identity manager
            expiresAt: expiresAt,
            groupId: groupId,
            coopId: coopId,
            contactMethod: contactMethod
        )

        try store.saveListing(listing)
        logger.info("Created listing: \(listing.id)")
        return listing
    }

    /// Register a co-op profile
    public func registerCoop(
        name: String,
        description: String?,
        memberCount: Int,
        governanceModel: GovernanceModel,
        industry: String,
        location: LocationValue?,
        website: String?,
        nostrPubkey: String,
        groupId: String?
    ) async throws -> CoopProfile {
        let coop = CoopProfile(
            name: name,
            description: description,
            memberCount: memberCount,
            governanceModel: governanceModel,
            industry: industry,
            location: location,
            website: website,
            nostrPubkey: nostrPubkey,
            groupId: groupId
        )

        try store.saveCoopProfile(coop)
        logger.info("Registered co-op: \(coop.id)")
        return coop
    }

    /// Get listings from the store
    public func getListings(
        groupId: String? = nil,
        type: ListingType? = nil,
        activeOnly: Bool = true
    ) throws -> [Listing] {
        try store.getListings(groupId: groupId, type: type, activeOnly: activeOnly)
    }

    /// Get co-op profiles from the store
    public func getCoopProfiles(groupId: String? = nil) throws -> [CoopProfile] {
        try store.getCoopProfiles(groupId: groupId)
    }

    // MARK: - Nostr Event Handlers (private)

    private func handleListingEvent(_ event: NostrEvent) async {
        logger.debug("Received listing event: \(event.id ?? "unknown")")
        // Parse and store the listing from the event content
    }

    private func handleCoopEvent(_ event: NostrEvent) async {
        logger.debug("Received co-op event: \(event.id ?? "unknown")")
    }

    private func handleReviewEvent(_ event: NostrEvent) async {
        logger.debug("Received review event: \(event.id ?? "unknown")")
    }

    private func handleSkillExchangeEvent(_ event: NostrEvent) async {
        logger.debug("Received skill exchange event: \(event.id ?? "unknown")")
    }

    private func handleResourceShareEvent(_ event: NostrEvent) async {
        logger.debug("Received resource share event: \(event.id ?? "unknown")")
    }
}

// MARK: - Nostr Event Kinds

/// Nostr event kinds for marketplace (40131-40139)
public enum MarketplaceNostrKinds {
    public static let listing = 40131
    public static let coopProfile = 40132
    public static let review = 40133
    public static let skillExchange = 40134
    public static let resourceShare = 40135
}

File: clients/android/app/src/main/java/network/buildit/modules/marketplace/MarketplaceModule.kt
package network.buildit.modules.marketplace

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Store
import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import dagger.multibindings.IntoSet
import kotlinx.serialization.json.Json
import network.buildit.core.modules.BuildItModule
import network.buildit.core.modules.ModuleRoute
import network.buildit.core.nostr.NostrClient
import network.buildit.core.nostr.NostrEvent
import network.buildit.core.storage.BuildItDatabase
import network.buildit.modules.marketplace.data.MarketplaceRepository
import network.buildit.modules.marketplace.data.local.CoopProfileEntity
import network.buildit.modules.marketplace.data.local.CoopProfilesDao
import network.buildit.modules.marketplace.data.local.GovernanceModel
import network.buildit.modules.marketplace.data.local.ListingEntity
import network.buildit.modules.marketplace.data.local.ListingStatus
import network.buildit.modules.marketplace.data.local.ListingType
import network.buildit.modules.marketplace.data.local.ListingsDao
import network.buildit.modules.marketplace.data.local.ResourceSharesDao
import network.buildit.modules.marketplace.data.local.ReviewsDao
import network.buildit.modules.marketplace.data.local.SkillExchangesDao
import network.buildit.modules.marketplace.domain.MarketplaceUseCase
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Marketplace module for BuildIt.
 *
 * Provides:
 * - Marketplace listings (products, services, co-ops, initiatives, resources)
 * - Co-op directory with governance models
 * - Skill exchange / time-banking
 * - Resource sharing (tools, spaces, vehicles)
 * - Community reviews
 */
class MarketplaceModule @Inject constructor(
    private val marketplaceUseCase: MarketplaceUseCase,
    private val repository: MarketplaceRepository,
    private val nostrClient: NostrClient
) : BuildItModule {
    override val identifier: String = "marketplace"
    override val version: String = "1.0.0"
    override val displayName: String = "Marketplace"
    override val description: String = "Cooperative marketplace with skill exchange and resource sharing"

    private var subscriptionId: String? = null

    override suspend fun initialize() {
        // Subscribe to marketplace-related Nostr events
        subscriptionId = nostrClient.subscribe(
            network.buildit.core.nostr.NostrFilter(
                kinds = getHandledEventKinds(),
                since = System.currentTimeMillis() / 1000 - 86400 * 7 // Last 7 days
            )
        )
    }

    override suspend fun shutdown() {
        subscriptionId?.let { nostrClient.unsubscribe(it) }
    }

    override suspend fun handleEvent(event: NostrEvent): Boolean {
        return when (event.kind) {
            MarketplaceUseCase.KIND_LISTING -> {
                handleListingEvent(event)
                true
            }
            MarketplaceUseCase.KIND_COOP_PROFILE -> {
                handleCoopProfileEvent(event)
                true
            }
            MarketplaceUseCase.KIND_REVIEW -> {
                handleReviewEvent(event)
                true
            }
            MarketplaceUseCase.KIND_SKILL_EXCHANGE -> {
                handleSkillExchangeEvent(event)
                true
            }
            MarketplaceUseCase.KIND_RESOURCE_SHARE -> {
                handleResourceShareEvent(event)
                true
            }
            NostrClient.KIND_DELETE -> {
                handleDeletionEvent(event)
                true
            }
            else -> false
        }
    }

    override fun getNavigationRoutes(): List<ModuleRoute> {
        return listOf(
            ModuleRoute(
                route = "marketplace",
                title = "Marketplace",
                icon = Icons.Default.Store,
                showInNavigation = true,
                content = { args ->
                    // MarketplaceScreen handled by navigation
                }
            ),
            ModuleRoute(
                route = "marketplace/{listingId}",
                title = "Listing Details",
                icon = null,
                showInNavigation = false,
                content = { args ->
                    val listingId = args["listingId"] ?: return@ModuleRoute
                    // ListingDetailScreen handled by navigation
                }
            ),
            ModuleRoute(
                route = "marketplace/create",
                title = "Create Listing",
                icon = null,
                showInNavigation = false,
                content = { args ->
                    // CreateListingScreen handled by navigation
                }
            )
        )
    }

    override fun getHandledEventKinds(): List<Int> {
        return listOf(
            MarketplaceUseCase.KIND_LISTING,
            MarketplaceUseCase.KIND_COOP_PROFILE,
            MarketplaceUseCase.KIND_REVIEW,
            MarketplaceUseCase.KIND_SKILL_EXCHANGE,
            MarketplaceUseCase.KIND_RESOURCE_SHARE,
            NostrClient.KIND_DELETE
        )
    }

    /**
     * Handles incoming listing events from Nostr.
     */
    private suspend fun handleListingEvent(nostrEvent: NostrEvent) {
        try {
            val json = Json { ignoreUnknownKeys = true }
            val content = json.decodeFromString<Map<String, kotlinx.serialization.json.JsonElement>>(nostrEvent.content)

            val listing = ListingEntity(
                id = content["id"]?.toString()?.trim('"') ?: return,
                schemaVersion = content["_v"]?.toString()?.trim('"') ?: "1.0.0",
                type = ListingType.fromValue(content["type"]?.toString()?.trim('"') ?: "product"),
                title = content["title"]?.toString()?.trim('"') ?: return,
                description = content["description"]?.toString()?.trim('"'),
                price = content["price"]?.toString()?.toDoubleOrNull(),
                currency = content["currency"]?.toString()?.trim('"') ?: "USD",
                imagesJson = content["images"]?.toString(),
                locationJson = content["location"]?.toString(),
                availability = content["availability"]?.toString()?.trim('"'),
                tagsJson = content["tags"]?.toString(),
                createdBy = content["createdBy"]?.toString()?.trim('"') ?: nostrEvent.pubkey,
                createdAt = content["createdAt"]?.toString()?.toLongOrNull() ?: nostrEvent.createdAt,
                updatedAt = content["updatedAt"]?.toString()?.toLongOrNull(),
                expiresAt = content["expiresAt"]?.toString()?.toLongOrNull(),
                status = ListingStatus.fromValue(content["status"]?.toString()?.trim('"') ?: "active"),
                groupId = nostrEvent.tags.find { it.firstOrNull() == "g" }?.getOrNull(1),
                coopId = nostrEvent.tags.find { it.firstOrNull() == "p" }?.getOrNull(1),
                contactMethod = content["contactMethod"]?.toString()?.trim('"') ?: "dm"
            )

            repository.saveListing(listing)
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle listing event", e)
        }
    }

    /**
     * Handles incoming co-op profile events from Nostr.
     */
    private suspend fun handleCoopProfileEvent(nostrEvent: NostrEvent) {
        try {
            val json = Json { ignoreUnknownKeys = true }
            val content = json.decodeFromString<Map<String, kotlinx.serialization.json.JsonElement>>(nostrEvent.content)

            val coop = CoopProfileEntity(
                id = content["id"]?.toString()?.trim('"') ?: return,
                schemaVersion = content["_v"]?.toString()?.trim('"') ?: "1.0.0",
                name = content["name"]?.toString()?.trim('"') ?: return,
                description = content["description"]?.toString()?.trim('"'),
                memberCount = content["memberCount"]?.toString()?.toIntOrNull() ?: 1,
                governanceModel = GovernanceModel.fromValue(
                    content["governanceModel"]?.toString()?.trim('"') ?: "consensus"
                ),
                industry = content["industry"]?.toString()?.trim('"') ?: "",
                locationJson = content["location"]?.toString(),
                website = content["website"]?.toString()?.trim('"'),
                nostrPubkey = content["nostrPubkey"]?.toString()?.trim('"') ?: nostrEvent.pubkey,
                verifiedByJson = content["verifiedBy"]?.toString(),
                image = content["image"]?.toString()?.trim('"'),
                createdAt = content["createdAt"]?.toString()?.toLongOrNull() ?: nostrEvent.createdAt,
                updatedAt = content["updatedAt"]?.toString()?.toLongOrNull(),
                groupId = nostrEvent.tags.find { it.firstOrNull() == "g" }?.getOrNull(1)
            )

            repository.saveCoopProfile(coop)
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle co-op profile event", e)
        }
    }

    /**
     * Handles incoming review events from Nostr.
     */
    private suspend fun handleReviewEvent(nostrEvent: NostrEvent) {
        try {
            val json = Json { ignoreUnknownKeys = true }
            val content = json.decodeFromString<Map<String, kotlinx.serialization.json.JsonElement>>(nostrEvent.content)

            val review = network.buildit.modules.marketplace.data.local.ReviewEntity(
                id = content["id"]?.toString()?.trim('"') ?: return,
                schemaVersion = content["_v"]?.toString()?.trim('"') ?: "1.0.0",
                listingId = content["listingId"]?.toString()?.trim('"') ?: return,
                reviewerPubkey = content["reviewerPubkey"]?.toString()?.trim('"') ?: nostrEvent.pubkey,
                rating = content["rating"]?.toString()?.toIntOrNull()?.coerceIn(1, 5) ?: return,
                text = content["text"]?.toString()?.trim('"') ?: return,
                createdAt = content["createdAt"]?.toString()?.toLongOrNull() ?: nostrEvent.createdAt
            )

            repository.saveReview(review)
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle review event", e)
        }
    }

    /**
     * Handles incoming skill exchange events from Nostr.
     */
    private suspend fun handleSkillExchangeEvent(nostrEvent: NostrEvent) {
        try {
            val json = Json { ignoreUnknownKeys = true }
            val content = json.decodeFromString<Map<String, kotlinx.serialization.json.JsonElement>>(nostrEvent.content)

            val exchange = network.buildit.modules.marketplace.data.local.SkillExchangeEntity(
                id = content["id"]?.toString()?.trim('"') ?: return,
                schemaVersion = content["_v"]?.toString()?.trim('"') ?: "1.0.0",
                offeredSkill = content["offeredSkill"]?.toString()?.trim('"') ?: return,
                requestedSkill = content["requestedSkill"]?.toString()?.trim('"') ?: return,
                availableHours = content["availableHours"]?.toString()?.toDoubleOrNull() ?: 0.0,
                hourlyTimebank = content["hourlyTimebank"]?.toString()?.toDoubleOrNull() ?: 0.0,
                locationJson = content["location"]?.toString(),
                createdBy = content["createdBy"]?.toString()?.trim('"') ?: nostrEvent.pubkey,
                createdAt = content["createdAt"]?.toString()?.toLongOrNull() ?: nostrEvent.createdAt,
                updatedAt = content["updatedAt"]?.toString()?.toLongOrNull(),
                status = network.buildit.modules.marketplace.data.local.SkillExchangeStatus.fromValue(
                    content["status"]?.toString()?.trim('"') ?: "active"
                ),
                groupId = nostrEvent.tags.find { it.firstOrNull() == "g" }?.getOrNull(1)
            )

            repository.saveExchange(exchange)
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle skill exchange event", e)
        }
    }

    /**
     * Handles incoming resource share events from Nostr.
     */
    private suspend fun handleResourceShareEvent(nostrEvent: NostrEvent) {
        try {
            val json = Json { ignoreUnknownKeys = true }
            val content = json.decodeFromString<Map<String, kotlinx.serialization.json.JsonElement>>(nostrEvent.content)

            val resource = network.buildit.modules.marketplace.data.local.ResourceShareEntity(
                id = content["id"]?.toString()?.trim('"') ?: return,
                schemaVersion = content["_v"]?.toString()?.trim('"') ?: "1.0.0",
                resourceType = network.buildit.modules.marketplace.data.local.ResourceShareType.fromValue(
                    content["resourceType"]?.toString()?.trim('"') ?: "tool"
                ),
                name = content["name"]?.toString()?.trim('"') ?: return,
                description = content["description"]?.toString()?.trim('"'),
                imagesJson = content["images"]?.toString(),
                locationJson = content["location"]?.toString(),
                depositRequired = content["depositRequired"]?.toString() == "true",
                depositAmount = content["depositAmount"]?.toString()?.toDoubleOrNull(),
                depositCurrency = content["depositCurrency"]?.toString()?.trim('"'),
                createdBy = content["createdBy"]?.toString()?.trim('"') ?: nostrEvent.pubkey,
                createdAt = content["createdAt"]?.toString()?.toLongOrNull() ?: nostrEvent.createdAt,
                updatedAt = content["updatedAt"]?.toString()?.toLongOrNull(),
                status = network.buildit.modules.marketplace.data.local.ResourceShareStatus.fromValue(
                    content["status"]?.toString()?.trim('"') ?: "available"
                ),
                groupId = nostrEvent.tags.find { it.firstOrNull() == "g" }?.getOrNull(1)
            )

            repository.saveResource(resource)
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle resource share event", e)
        }
    }

    /**
     * Handles deletion events from Nostr.
     */
    private suspend fun handleDeletionEvent(nostrEvent: NostrEvent) {
        try {
            val entityIds = nostrEvent.tags.filter { it.firstOrNull() == "e" }
                .mapNotNull { it.getOrNull(1) }

            entityIds.forEach { entityId ->
                // Try to delete from all tables (only one will match)
                repository.deleteListing(entityId)
                repository.deleteCoopProfile(entityId)
                repository.deleteReview(entityId)
                repository.deleteExchange(entityId)
                repository.deleteResource(entityId)
            }
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle deletion event", e)
        }
    }

    companion object {
        private const val TAG = "MarketplaceModule"
    }
}

/**
 * Hilt module for Marketplace dependencies.
 */
@Module
@InstallIn(SingletonComponent::class)
abstract class MarketplaceHiltModule {
    @Binds
    @IntoSet
    abstract fun bindMarketplaceModule(impl: MarketplaceModule): BuildItModule
}

/**
 * Provides DAOs for Marketplace module.
 */
@Module
@InstallIn(SingletonComponent::class)
object MarketplaceDaoModule {
    @Provides
    @Singleton
    fun provideListingsDao(database: BuildItDatabase): ListingsDao = database.listingsDao()

    @Provides
    @Singleton
    fun provideCoopProfilesDao(database: BuildItDatabase): CoopProfilesDao = database.coopProfilesDao()

    @Provides
    @Singleton
    fun provideReviewsDao(database: BuildItDatabase): ReviewsDao = database.reviewsDao()

    @Provides
    @Singleton
    fun provideSkillExchangesDao(database: BuildItDatabase): SkillExchangesDao = database.skillExchangesDao()

    @Provides
    @Singleton
    fun provideResourceSharesDao(database: BuildItDatabase): ResourceSharesDao = database.resourceSharesDao()
}
